File: 048/eu048.go
// Project Euler problem 48
// Solution by Kevin Retzke, June 2012
package main

import (
	"fmt"
)

// TruncInt stores a truncated integer.
type TruncInt struct {
	Digits    []int
	Precision int
}

// Print TruncInt value.
func (t *TruncInt) PrintValue() {
	for i := t.Precision - 1; i >= 0; i-- {
		fmt.Printf("%d", t.Digits[i])
	}
	fmt.Println()
}

// Init initializes a TruncInt type.
func (t *TruncInt) Init(value, precision int) {
	t.Digits = make([]int, precision)
	t.Precision = precision
	for i := 0; i < precision; i++ {
		t.Digits[i] = value % 10
		value = value / 10
	}
}

// Add o to t, results in t.
func (t *TruncInt) Add(o *TruncInt) {
	precision := t.Precision
	if o.Precision < precision {
		precision = o.Precision
	}
	for i := 0; i < precision-1; i++ {
		t.Digits[i] = t.Digits[i] + o.Digits[i]
		t.Digits[i+1] += t.Digits[i] / 10
		t.Digits[i] = t.Digits[i] % 10
	}
	if precision < t.Precision {
		t.Digits[precision] = t.Digits[precision-1] / 10
	}
	t.Digits[precision-1] = (t.Digits[precision-1] + o.Digits[precision-1]) % 10
}

// powmod computes the last digits of base**expo
func powmod(base, expo, digits int) TruncInt {
	var base_t, res_t TruncInt
	base_t.Init(base, digits)
	res_t.Init(base, digits)
	// expand base**expo into base*base*base*...
	for e := 1; e < expo; e++ {
		var tmp_t TruncInt
		tmp_t.Init(0, digits)
		// perform long multiplication
		for i := 0; i < digits; i++ {
			for j := 0; j < digits-i; j++ {
				tmp_t.Digits[i+j] += res_t.Digits[j] * base_t.Digits[i]
			}
		}
		// carry and copy to result array
		for i := 0; i < digits-1; i++ {
			tmp_t.Digits[i+1] += tmp_t.Digits[i] / 10
			res_t.Digits[i] = tmp_t.Digits[i] % 10
		}
		res_t.Digits[digits-1] = tmp_t.Digits[digits-1] % 10
	}
	return res_t
}

func main() {
	res_digits := 10
	var res TruncInt
	res.Init(0, res_digits)
	for i := 1; i < 1000; i++ {
		tmp := powmod(i, i, res_digits)
		res.Add(&tmp)
	}
	res.PrintValue()
}

File: c/eu048.c
// Project Euler problem 48
// Solution by Kevin Retzke, June 2012
#include <stdio.h>
#include <stdlib.h>

typedef unsigned int uint;
typedef unsigned long ulong;
typedef unsigned int byte;

// print number array
void printArray(byte *array, size_t len)
{
	for (int i = len-1; i >= 0; i--) {
		printf("%d",array[i]);
	}
}


// expand number to array, e.g. 123 -> {3,2,1}
void makeArray(uint num, byte *array, size_t len)
{
	uint rem = num;
	for (uint i=0; i<len; i++) {
		array[i] = rem%10;
		rem = rem/10;
	}
}

// add array2 to array1, results in array1
void addArray(byte *array1, const byte *array2, size_t len)
{
	for (uint i=0; i<len-1; i++) {
		array1[i] = array1[i]+array2[i];
		array1[i+1] += array1[i]/10;
		array1[i] = array1[i]%10;
	}
	array1[len-1] = (array1[len-1]+array2[len-1])%10;
}

// powmod computes the last n digits of base**expo
void powmod(uint base, uint expo, byte *res_array, size_t digits)
{
	byte base_array[digits];
	makeArray(base, base_array, digits);
	makeArray(base, res_array, digits);
	// expand base**expo into base*base*base*...
	for (uint e = 1; e < expo; e++) {
		byte tmp_array[digits];
		makeArray(0, tmp_array, digits);
		// perform long multiplication
		for (uint i = 0; i < digits; i++) {
			for (uint j = 0; j < digits-i; j++) {
				tmp_array[i+j] += res_array[j]*base_array[i];
			}
		}
		// carry and copy to result array
		for (uint i = 0; i < digits-1; i++) {
			tmp_array[i+1] += tmp_array[i]/10;
			res_array[i] = tmp_array[i]%10;
		}
		res_array[digits-1] = tmp_array[digits-1]%10;
	}
}

int main()
{
	static const size_t res_digits = 10;
	byte res_array[res_digits], tmp_array[res_digits];
	makeArray(0,res_array,res_digits);
	for (uint i=1; i<=1000; i++) {
		powmod(i,i,tmp_array,res_digits);
		addArray(res_array,tmp_array,res_digits);
		/*printf("%4d\t",i);
		printArray(tmp_array, res_digits);
		printf("\t");
		printArray(res_array, res_digits);
		printf("\n");*/
	}
	//powmod(999,999,res_array,res_digits);
	printArray(res_array, res_digits);
	printf("\n");
	return 0;
}
